streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -16,44 +16,59 @@
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.KafkaFuture;
 import org.apache.kafka.common.utils.LogContext;
-import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.StreamsConfig;
 import org.apache.kafka.streams.errors.StreamsException;
 import org.slf4j.Logger;
 
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Properties;
 import java.util.Set;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
 public class InternalTopicManager {
 
-    static final Long WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT = TimeUnit.MILLISECONDS.convert(1, TimeUnit.DAYS);
+    private static final Long WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT = TimeUnit.MILLISECONDS.convert(1, TimeUnit.DAYS);
 
     public static final String CLEANUP_POLICY_PROP = "cleanup.policy";
     public static final String RETENTION_MS = "retention.ms";
-    private static final int MAX_TOPIC_READY_TRY = 5;
 
     private final Logger log;
-    private final Time time;
     private final long windowChangeLogAdditionalRetention;
+    private final Map<String, String> defaultTopicConfigs = new HashMap<>();
 
-    private final int replicationFactor;
-    private final StreamsKafkaClient streamsKafkaClient;
+    private final short replicationFactor;
+    private final AdminClient adminClient;
 
-    public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
-                                final int replicationFactor,
-                                final long windowChangeLogAdditionalRetention,
-                                final Time time) {
-        this.streamsKafkaClient = streamsKafkaClient;
-        this.replicationFactor = replicationFactor;
-        this.windowChangeLogAdditionalRetention = windowChangeLogAdditionalRetention;
-        this.time = time;
+    public InternalTopicManager(final AdminClient adminClient,
+                                final Map<String, ?> config) {
+        this.adminClient = adminClient;
+        this.replicationFactor =  (short) (config.containsKey(StreamsConfig.REPLICATION_FACTOR_CONFIG) ? (Integer) config.get(StreamsConfig.REPLICATION_FACTOR_CONFIG) : 1);
+        this.windowChangeLogAdditionalRetention = config.containsKey(StreamsConfig.WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG) ?
+            (Long) config.get(StreamsConfig.WINDOW_STORE_CHANGE_LOG_ADDITIONAL_RETENTION_MS_CONFIG)
+            : WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
 
         LogContext logContext = new LogContext(String.format("stream-thread [%s] ", Thread.currentThread().getName()));
         this.log = logContext.logger(getClass());
+
+        for (final Map.Entry<String, Object> entry : new AdminClientConfig(config).originalsWithPrefix(StreamsConfig.TOPIC_PREFIX).entrySet()) {
+            if (entry.getValue() != null) {
```

Comment:
Could we ever have null values?

Reply:
Yes. If user explicitly sets a config to `null` -- very rare case but would be possible.